includes/class-events-store.php

```diff
@@ -882,6 +879,315 @@ public function count_events_by_status( $status ) {
 
 		return (int) $wpdb->get_var( $wpdb->prepare( "SELECT COUNT(ID) FROM {$this->get_table_name()} WHERE status = %s", $status ) ); // Cannot prepare table name. @codingStandardsIgnoreLine
 	}
+
+	/*
+	|--------------------------------------------------------------------------
+	| New event's store methods. The above may be deprecated in the future.
+	| But notably, the below is also internal-usage only. See comments about alternatives.
+	|--------------------------------------------------------------------------
+	*/
+
+	/**
+	 * Create an event.
+	 * For internal use only, please use Event:save() as this method does not validate.
+	 *
+	 * @param array $row_data The row data used to create the event.
+	 * @return int The newly created event ID, 0 if creation failed.
+	 */
+	public static function _create_event( array $row_data ): int {
+		global $wpdb;
+
+		if ( empty( $row_data ) ) {
+			return 0;
+		}
+
+		$row_data['created']       = current_time( 'mysql', true );
+		$row_data['last_modified'] = current_time( 'mysql', true );
+
+		$result = $wpdb->insert( self::table_name(), $row_data, self::row_formatting( $row_data ) );
+
+		self::flush_event_cache();
+		return false === $result ? 0 : $wpdb->insert_id;
+	}
+
+	/**
+	 * Update an event.
+	 * For internal use only, please use Event::save() as this does not validate.
+	 *
+	 * @param int   $event_id The ID of the event being updated.
+	 * @param array $row_data The row data used to update the event.
+	 * @return bool True if update was successful, false otherwise.
+	 */
+	public static function _update_event( int $event_id, array $row_data ): bool {
```

Comment:
Same as above... 

or if we want to abstract away DB operations from event class. Let's make a new class like `EventRepository` and inject it to Event class. 

That might actually work even better and decouple dependencies.

You could have `Event( $eventRepository )` then:
```
class EventFactory {
  EventRepository $eventRepository
  
  __constructor() {
     $eventRepository = new EventRepository();
  }

   public createEvent() {
      return new Event($eventRepository);
   }
   
  public getFromEventData($eventData) {
      $event = $this->createEvent();
      $event->hidrate($eventData);
      return $event;
   }
   
   public getFromId($id) {
      $eventData = $eventRepository->get($id);
      return $this->getFromEventData($eventData);
   }
}
```

The benefit is that it is all decoupled. You can test the repository without Event object. And event object with the mocked repository.

We can of course name `EventFactory` `EventStore` - that is this class, if that makes more sense.

Reply:
A few things come to mind:

1) It is a bit heavy on the added complexity side, though there are benefits of course.
2) Real unit testing is made easier w/ the factory, but WP Unit tests are more-so integration tests which is fine but as a result we don't get much benefit since we might as well test the full route.
3) It would be initializing a lot more objects when dealing w/ thousands of `Event`s, though this is probably a non-issue.
4) Not totally sure where things like `STATUS_PENDING` const would belong in the new structure long-term.

Was thinking maybe could cut out the factory middleman initially, and go w/:

```
class Event {
  private EventRepository $eventRepository

  __constructor() {
     $this->eventRepository = new EventRepository();
  }

   public function save() {
      ... logic ...
      return $this->eventRepository->update( $args );
   }
}
```

It keeps the logic enclosed so we'd be free to change it. And if in the future we ever wanted to support custom data stores, then adding the factory layer could make sense - providing a filter essentially to swap out the store for anther that provides the same interface.
____

But even w/ the above, I think we still end up with the same problem don't we? `EventRepository` is effectively what these three "fake private" methods are in Events_Store right now, just w/ extra steps. They have to live somewhere together, and they'd ultimately be accessible?

Reply:
Yeah I like that... the injecting in the constructor is what I was after. `Factory` is an afterthought. 

> EventRepository is effectively what these three "fake private" methods are in Events_Store right now

I think in this scenario EventRepository is EventStore so no extra step. Also, I think it is perfectly fine to have them public then. This fake_privatenes  + the fact that they are only used at one place (so dont need to be public if we would like to enforce them being private) is what confused me in the first place. :) 

I mean it is fine to expose the registry functionality so that people dont have to write their own SQL if they would like to skip the Event abstraction.

Thinking about it some more... maybe the issue here is that we are mixing the abstraction layer that the user should interact with (the EventStore/Event) together with the lower level Data Access layer that we don't want the user to interact with. Making sort of cyclic dependency. `EventStore` -> `Event` -> `EventStore (for SQL stuff)`. If we move out the SQL stuff to a separate class it might be cleaner `EventStore` -> `Event` -> `EventRepository` and `EventStore` -> `EventRepositoy`.

What do you think?

Reply:
Summarizing our chat for historical record:

- There won't be circular dependencies in the end result, it will go just from Event or Events -> EventsStore. In the interim there may be some circles while we deprecate methods in the store, but the end result will be linear. The store is the last stop, handling the SQL and caching layers.

- As such, the end result is that the consumer/user will never be calling the EventStore methods at all once everything is in place. Events and Event will hold wrappers for any SQL-related functionality so we can ensure it goes through the right paths.